src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,223 @@
+                entity.setProperty("rowNum", location.row);
+                entity.setProperty("colNum", location.col);
+
+                // Write to datastore.
+                datastore.put(entity);
+            }
+        }
+    }
+
+    /**
+     * Read from RealBoard and construct a Json string encoding the board state.
+     * @param simulationId is the simulation id to be retrieved.
+     * @param round is the round to be retrieved.
+     * @return a Json string encoding the board state, uses for visualization,
+     * or null if provided round exceeds simulation's maximum number of rounds.
+     */
+    public String readRealBoardState(String simulationId, int round) {
+        return readBoardState(simulationId, round, "RealBoard");
+    }
+
+    /**
+     * Read from EstimatedBoard and construct a Json string encoding the board state.
+     * @param simulationId is the simulation id to be retrieved.
+     * @param round is the round to be retrieved.
+     * @return a Json string encoding the board state, uses for visualization,
+     * or null if provided round exceeds simulation's maximum number of rounds.
+     */
+    public String readEstimatedBoardState(String simulationId, int round) {
+        return readBoardState(simulationId, round, "EstimatedBoard");
+    }
+
+    private String readBoardState(String simulationId, int round, String entityKind) {
+        // throw an exception if provided round exceeds simulation's maximum number of rounds.
+        if (!isRoundExistsInSimulation(simulationId, round)) {
+            throw new ExceedingRoundException("Provided round " + round +
+                    " exceeds maximum number of rounds of simulation " + simulationId);
+        }
+
+        // Set simple predicates.
+        Query.FilterPredicate filterBySimulationId =
+                new Query.FilterPredicate("simulationId", Query.FilterOperator.EQUAL, simulationId);
+        Query.FilterPredicate filterByRound =
+                new Query.FilterPredicate("round", Query.FilterOperator.EQUAL, round);
+
+        // Compose simple predicates.
+        Query.CompositeFilter composedFilter = Query.CompositeFilterOperator.and(filterBySimulationId, filterByRound);
+
+        // Create query and return result.
+        Query boardStateBySimulationIdAndRoundQuery = new Query(entityKind).setFilter(composedFilter);
+        PreparedQuery boardStateBySimulationIdAndRoundPreparedQuery = datastore.prepare(boardStateBySimulationIdAndRoundQuery);
+
+        // Return encoded board state.
+        return toJsonTable(boardStateBySimulationIdAndRoundPreparedQuery.asIterable(), simulationId);
+    }
+
+    private String toJsonTable(Iterable<Entity> boardStateEntities, String simulationId) {
+        // Retrieve simulations' dimensions.
+        Entity simulationEntity = retrieveSimulationEntity(simulationId);
+        int rowNum = ((Long) simulationEntity.getProperty("rowNum")).intValue();
```

Comment:
this highly couples the database utility with the simulation implementation/interface

Reply:
Here the database utility is dependent on the datastore Simulation entity structure (i.e its schema), not the simulation implementation/interface. How can we avoid this? when reading an entity from datastore (or any other relational/non-relational db) we must specify the entity's properties names.

@schainmariano